lib/SPIRV/PreprocessMetadata.cpp

```diff
@@ -98,50 +95,55 @@ bool TransOCLMD::runOnModule(Module &Module) {
   return true;
 }
 
-void TransOCLMD::visit(Module *M) {
+void PreprocessMetadata::visit(Module *M) {
   SPIRVMDBuilder B(*M);
   SPIRVMDWalker W(*M);
-  // !spirv.Source = !{!x}
-  // !{x} = !{i32 3, i32 102000}
-  B.addNamedMD(kSPIRVMD::Source)
-      .addOp()
-      .add(CLVer < kOCLVer::CL21 ? spv::SourceLanguageOpenCL_C
-                                 : spv::SourceLanguageOpenCL_CPP)
-      .add(CLVer)
-      .done();
-  if (EraseOCLMD)
-    B.eraseNamedMD(kSPIR2MD::OCLVer).eraseNamedMD(kSPIR2MD::SPIRVer);
-
-  // !spirv.MemoryModel = !{!x}
-  // !{x} = !{i32 1, i32 2}
-  Triple TT(M->getTargetTriple());
-  assert(isSupportedTriple(TT) && "Invalid triple");
-  B.addNamedMD(kSPIRVMD::MemoryModel)
-      .addOp()
-      .add(TT.isArch32Bit() ? spv::AddressingModelPhysical32
-                            : spv::AddressingModelPhysical64)
-      .add(spv::MemoryModelOpenCL)
-      .done();
-
-  // Add source extensions
-  // !spirv.SourceExtension = !{!x, !y, ...}
-  // !x = {!"cl_khr_..."}
-  // !y = {!"cl_khr_..."}
-  auto Exts = getNamedMDAsStringSet(M, kSPIR2MD::Extensions);
-  if (!Exts.empty()) {
-    auto N = B.addNamedMD(kSPIRVMD::SourceExtension);
-    for (auto &I : Exts)
-      N.addOp().add(I).done();
+
+  unsigned CLVer = getOCLVersion(M, true);
+  if (CLVer != 0) { // Preprocess OpenCL-specific metadata
+    // !spirv.Source = !{!x}
+    // !{x} = !{i32 3, i32 102000}
+    B.addNamedMD(kSPIRVMD::Source)
+        .addOp()
+        .add(CLVer < kOCLVer::CL21 ? spv::SourceLanguageOpenCL_C
+                                   : spv::SourceLanguageOpenCL_CPP)
+        .add(CLVer)
+        .done();
+    if (EraseOCLMD)
+      B.eraseNamedMD(kSPIR2MD::OCLVer).eraseNamedMD(kSPIR2MD::SPIRVer);
+
+    // !spirv.MemoryModel = !{!x}
+    // !{x} = !{i32 1, i32 2}
+    Triple TT(M->getTargetTriple());
+    assert(isSupportedTriple(TT) && "Invalid triple");
+    B.addNamedMD(kSPIRVMD::MemoryModel)
```

Comment:
> Main purpose of this PR is to allow SPIRV-LLVM-Translator to handle some OpenCL-specific metadata (like reqd_work_group_size, intel_reqd_sub_group_size) even if LLVM IR wasn't generated from OpenCL C (opencl.ocl.version metadata is missed).


Can we also untie setting of addressing model from OpenCL version metadata presence?

Just out of the curiosity: do you know why we set AddressingModel based on the triple and not data layout?

Reply:
Not sure that it is a good idea. SPIR-V spec defines [two more AddressingModels](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_addressing_model_a_addressing_model) and I understand  we need to have some default value, but we already have such:
* [Dummy initial value in SPIRVModuleImpl](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/b46b910eeaee4ac4a8e2360a9240d68015dc3a59/lib/SPIRV/libSPIRV/SPIRVModule.cpp#L68)
* [transAddressingMode](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/1d9349cf53a47065ade67d86d687a78e4c5c29a4/lib/SPIRV/SPIRVWriter.cpp#L1408) which is used [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/1d9349cf53a47065ade67d86d687a78e4c5c29a4/lib/SPIRV/SPIRVWriter.cpp#L1589)

I just don't want to add another one default value which needs to be changed later if we decide to support different flavors of SPIR-V (#30).

Moreover, looks like this MemoryModel spirv-friendly metadata is also unused. BTW, AddressingModel is stored together with MemoryModel in spirv-friendly metadata, so I cannot just translate one of them without additional changes.